app/models/miq_server/worker_management/monitor.rb

```diff
@@ -40,21 +40,14 @@ def worker_not_responding(w)
   end
 
   def sync_workers
-    result = {}
-    MiqWorkerType.worker_class_names.each do |class_name|
-      begin
-        c = class_name.constantize
-        raise NameError, "Constant problem: expected: #{class_name}, constantized: #{c.name}" unless c.name == class_name
-
-        result[c.name] = c.sync_workers
-        result[c.name][:adds].each { |pid| worker_add(pid) unless pid.nil? }
-      rescue => error
-        _log.error("Failed to sync_workers for class: #{class_name}")
-        _log.log_backtrace(error)
-        next
-      end
+    MiqWorkerType.worker_class_names.map(&:constantize).each_with_object({}) do |klass, result|
```

Comment:
Wondering if `MiqWorkerType.worker_class_names.map(&:constantize)` should be extracted to a method like `MiqWorkerType.worker_classes`

Reply:
:100: may or may not already have that branch as a follow-up to this :)

Reply:
doing a search in the core repo shows
```sh
$ git grep worker_class_names
app/models/miq_server/worker_management/monitor.rb:44:    MiqWorkerType.worker_class_names.each do |class_name|
app/models/miq_server/worker_management/monitor.rb:163:    MiqWorkerType.worker_class_names_in_kill_order.each do |class_name|
app/models/miq_server/worker_management/monitor/settings.rb:11:    MiqWorkerType.worker_class_names.each do |class_name|
app/models/miq_server/worker_management/monitor/start.rb:13:      starting = MiqWorker.find_starting.find_all { |w| MiqWorkerType.worker_class_names.include?(w.class.name)
app/models/miq_server/worker_management/monitor/systemd.rb:51:      MiqWorkerType.worker_class_names.map(&:constantize).map(&:service_base_name)
app/models/miq_worker_type.rb:27:  def self.worker_class_names
app/models/miq_worker_type.rb:31:  def self.worker_class_names_in_kill_order
spec/models/miq_server/worker_management/monitor_spec.rb:99:        allow(MiqWorkerType).to receive(:worker_class_names).and_return(%w[ManageIQ::Providers::Foreman::Provis
spec/models/miq_server/worker_management/monitor_spec.rb:107:        allow(MiqWorkerType).to receive(:worker_class_names).and_return(%w(MiqGenericWorker MiqPriorityWorker)
spec/models/miq_server/worker_management/monitor_spec.rb:114:        allow(MiqWorkerType).to receive(:worker_class_names).and_return([])
spec/models/miq_worker/systemd_common_spec.rb:10:      found_units = MiqWorkerType.worker_class_names.flat_map do |klass_name|
```

I would bet that each of those callers is doing a `constantize` to get the actual classes.

Reply:
Yes not all of them are but the majority

Reply:
Note, `worker_classes` could add that `constantize` check on the class name matching the worker type name if that is really still needed.  Looking at my original PR, it looks like the problem was in `const_get` doing the wrong thing.  I guess that could still happen if rails `autoload` still does that.  🤷 

I can look to see if that's still an issue.    But here's more context: https://github.com/ManageIQ/manageiq/pull/19400#issuecomment-544536976

Reply:
Right AFAIK constantizing classes which don't exist won't always fail (could resolve to the non-namespaced base class), but constantizing classes which do exist will always resolve to the right thing?

Reply:
@agrare I believe that's correct.

Reply:
Note that newer versions of Ruby / Rails may have eliminated this problem entirely as well, but in this case you know all the values will exist (because they were derived from existing values), so you are safe.